src/util/amount.rs

```diff
@@ -849,6 +856,42 @@ impl FromStr for SignedAmount {
     }
 }
 
+impl ::std::iter::Sum for SignedAmount {
+    fn sum<I: Iterator<Item=Self>>(iter: I) -> Self {
+        let sats: i64 = iter.map(|amt| amt.0).sum();
+        SignedAmount::from_sat(sats)
+    }
+}
+
+/// Calculate the sum over the iterator using checked arithmetic.
+pub trait CheckedSum<R> {
+    /// Calculate the sum over the iterator using checked arithmetic. If an over or underflow would
+    /// happen it returns `None`.
+    fn checked_sum(self) -> Option<R>;
+}
+
+impl<T> CheckedSum<Amount> for T where T: Iterator<Item = Amount> {
+    fn checked_sum(mut self) -> Option<Amount> {
+        let first = Some(self.next().unwrap_or_default());
+
+        self.fold(
+            first,
+            |acc, item| acc.and_then(|acc| acc.checked_add(item))
```

Comment:
I don't really like the code duplication, but the rest of the module doesn't use macros either to prevent it, so I followed this approach.